### PR TITLE
BUG: fix byte order reversal for datetime64[ns]

### PR DESCRIPTION
**Pull request by rharriszzz**
Numpy arrays with metadata (including units for datetime64) can't have their type copied based only on the typenum.
This change is in the array_setstate, where it does byte reversal of data from platforms with a different endian value.

### CODE REVIEW

**Comment by eric-wieser** [comment]
Can you:
* Adjust your test to call `__setstate__` and `__getstate__` via pickle?
* Commit the test within the PR?

**Comment by eric-wieser** [comment]
(closing and reopening to restart CI, since we had Azure problems until recently)

**Comment by rharriszzz** [comment]
I committed the test program.  I suppose the test could be moved into test_multiarray.py, if you want.

File: numpy/core/tests/test_numpy_12905.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    original_byte_reversed = original.copy(order='K')
+    original_byte_reversed.dtype = original_byte_reversed.dtype.newbyteorder('S')
+    original_byte_reversed.byteswap(inplace=True)
+
+    new = pickle.loads(pickle.dumps(original_byte_reversed))
+    
+    assert_equal(original.dtype, new.dtype)
+
+if __name__ == "__main__":
+    run_module_suite()
```
Comment (eric-wieser) [inline numpy/core/tests/test_numpy_12905.py]: nit: I think we tend to skip this now.

File: numpy/core/tests/test_numpy_12905.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+
+import sys
+import pickle
+
+import numpy as np
+from numpy.testing import (
+     run_module_suite, assert_, assert_equal, 
+)
+
+def test_numpy_12905():
```
Comment (eric-wieser) [inline numpy/core/tests/test_numpy_12905.py]: I think this should move to `numpy\core\tests\test_multiarray.py:TestPickling.test_datetime64_byteorder

**Comment by eric-wieser** [comment]
>  I suppose the test could be moved into test_multiarray.py, if you want.

Yes please - I added a comment above describing where I think it should go.

Sorry for leaving you hanging here, everything else looks good

File: numpy/core/tests/test_numpy_12905.py (hunk 2 above)
Comment (rharriszzz) [reply numpy/core/tests/test_numpy_12905.py]: Done.

File: numpy/core/src/multiarray/methods.c [hunk 3]
```diff
@@ -2041,7 +2041,9 @@ array_setstate(PyArrayObject *self, PyObject *args)
                                         PyArray_DESCR(self)->elsize,
                                         datastr, PyArray_DESCR(self)->elsize,
                                         numels, 1, self);
-                if (!PyArray_ISEXTENDED(self)) {
+                if (!(PyArray_ISEXTENDED(self) ||
+                      PyArray_DESCR(self)->metadata ||
+                      PyArray_DESCR(self)->c_metadata)) {
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/methods.c]: What would go wrong if we removed this `if` entirely, I wonder?

**Comment by eric-wieser** [comment]
Thanks for your first contribution, @rharriszzz!

**Pull request merged**